PSA: Phase 1: Test using opaque keys for CSR creation

AndrzejKurek opened the pull request:
This PR handles and tests CSR creation using opaque keys. It's described here: https://jira.arm.com/browse/IOTSSL-2575.

Note: this PR has to be reworked to use the common shared utilities.

__Based on:__ https://github.com/ARMmbed/mbedtls/pull/2162. Only [this commit](https://github.com/ARMmbed/mbedtls/commit/a34baf19b9a7e343569a33035d91270b8f02fb55) should be reviewed.

**Note:** will need rebasing on the submodule work before merging! The "DO NOT MERGE" label can be removed once it's done.

AndrzejKurek commented on library/pk_wrap.c:
```diff
@@ -716,4 +720,288 @@ const mbedtls_pk_info_t mbedtls_rsa_alt_info = {
+    mbedtls_asn1_buf signature;
+    psa_algorithm_t psa_sig_md = PSA_ALG_ECDSA( translate_md_to_psa( md_alg ) );
+    psa_key_slot_t* key_slot = (psa_key_slot_t *) ctx;
+
```
Looking at it now, I should probably add a check for key type to let only ECDSA pass.

AndrzejKurek commented:
Thanks for adjusting the name, @mpg! 🥇

hanno-arm commented on tests/suites/test_suite_x509write.function:
```diff
@@ -95,6 +167,48 @@ exit:
 /* END_CASE */
 
+/* BEGIN_CASE depends_on:MBEDTLS_PEM_WRITE_C:MBEDTLS_X509_CSR_WRITE_C:MBEDTLS_USE_PSA_CRYPTO */
+void x509_csr_check_opaque( char * key_file, int md_type, int key_usage,
```
Minor: Please correct the indentation

AndrzejKurek replied:
Done.

hanno-arm commented on tests/suites/test_suite_x509write.function:
```diff
@@ -95,6 +167,48 @@ exit:
+
+    mbedtls_pk_init( &key );
+    TEST_ASSERT( mbedtls_pk_parse_keyfile( &key, key_file, NULL ) == 0 );
+    TEST_ASSERT( wrap_key_as_opaque( &key, PSA_ALG_SHA_1 ) == 0 );
```
We shouldn't introduce new code building on SHA-1. Can you switch to SHA-256?

AndrzejKurek replied:
Done.

hanno-arm commented on tests/suites/test_suite_x509write.function:
```diff
@@ -95,6 +167,48 @@ exit:
+
+    pem_len = strlen( (char *) buf );
+    buf[pem_len] = '\0';
+    TEST_ASSERT( x509_crt_verifycsr(buf, pem_len+1, &key) == 0 );
```
Minor: Spacing

AndrzejKurek replied:
Done!

hanno-arm commented on tests/suites/test_suite_x509write.function:
```diff
@@ -28,6 +32,74 @@ size_t mbedtls_rsa_key_len_func( void *ctx )
 #endif /* MBEDTLS_RSA_C */
 
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+static int wrap_key_as_opaque( mbedtls_pk_context *pk, psa_algorithm_t hash_alg )
```
This is replicated from #2183. As mentioned in https://github.com/ARMmbed/mbedtls/pull/2183/files#r234435165, I think we should introduce it as shared code in `pk_internal.h` in #2162. @mpg @AndrzejKurek Please let me know what you think.

hanno-arm commented on tests/suites/test_suite_x509write.function:
```diff
@@ -95,6 +167,48 @@ exit:
+
+    pem_len = strlen( (char *) buf );
+    buf[pem_len] = '\0';
+    TEST_ASSERT( x509_crt_verifycsr(buf, pem_len+1, &key) == 0 );
```
Major: Why does this test deviate from the existing CSR writing tests in that it attempts a verification instead of a binary `memcmp()`? This (a) introduces the need to support opaque verification here which could otherwise be omitted for the benefit of simplicity of this PR, and (b) Introduces a dependency on X.509 parsing which doesn't exist anywhere else in the test suite.

@mpg @AndrzejKurek I'd suggest to adapt the code from https://github.com/ARMmbed/mbedtls/blob/development/tests/suites/test_suite_x509write.function#L39 to allow an additional `use_psa` parameter, calling `wrap_key_as_opaque()` if set, but otherwise keeping the code. All tests for `x509_csr_check()` can then be duplicated once with the new `use_psa` parameter set.

hanno-arm replied:
On a general note, could we please at least briefly document such design considerations and discussions somewhere? It makes review so much easier if one sees not only what has been done but also why, and in particular what has been considered and deliberately refused as an alternative, especially if it's the canonical approach.

Or is it documented somewhere and I just missed it again?

mpg replied:
> And why don't we have these problems for the ordinary CSR tests, too?

For the ordinary CRS test, we pass a `f_rng` argument to `mbedtls_x509write_csr_pem()` that's fully deterministic, and consider it enough to make the result fully deterministic. IMO that's fragile (imagine we change how we use the RNG for blinding or whatever in a way that affects what part of the RNG output is used for making the signature, or that we switch for deterministic ECDSA by default, or...) and I would do it differently if I was writing this test today, regardless of the next consideration.

The reason why this won't work with PSA is that it disregards `f_rng` and always uses its own. Getting to the internal PSA RNG and replacing it with a fixed one is technically possible but IMO poor design as it violates an abstract, in addition to the considerations above.

> Sorry for my ignorance, but why is the TBS part of the CSR random? 

The CSR includes a signature and we clearly want the signature to be verified, and not just the part being signed, as the signature is what we're modifying here. With ECDSA, the signature may be random. We shouldn't be making assumptions about whether the particular PSA implementation used does deterministic ECDSA by default or not.

> On a general note, could we please at least briefly document such design considerations and discussions somewhere? It makes review so much easier if one sees not only what has been done but also why, and in particular what has been considered and deliberately refused as an alternative, especially if it's the canonical approach.

I agree that kind of thing should be mentioned at least in the PR description, sometimes even in the commit message.

hanno-arm replied:
@mpg Thanks for your explanations. I initially thought you'd say that the to-be-signed part itself was random, but apparently that was a misunderstanding.

AndrzejKurek replied:
I'm sorry for not describing it properly - probably because I wanted the PR to be set up as fast as possible. I'll add that to the description. Thanks @mpg for explaining!

mpg commented on tests/suites/test_suite_x509write.function:
```diff
@@ -28,6 +32,74 @@ size_t mbedtls_rsa_key_len_func( void *ctx )
+        return( MBEDTLS_ERR_PK_TYPE_MISMATCH );
+
+    ec = mbedtls_pk_ec( *pk );
+    if( ( key_len = mbedtls_pk_write_key_der( pk, buf, buf_len ) ) <= 0 )
```
If `psa_import_key()` work after that, it means you're using the old version of that API. I'l ask Hanno to pick up my version, then you can rebase over the utilities branch and us the version from there.

mpg commented:
@AndrzejKurek #2162 is now ready to serve as a new base for your work here. I extracted the utility function that you'll also need here from the other PR.

AndrzejKurek commented:
@mpg, @hanno-arm - I've rebased the PR, simplified it and add a few tweaks to make it work better. I hope you don't mind the history rewriting, I figured it'll be more clear this way. Please re-review (only the last two commits).

hanno-arm commented on library/pkwrite.c:
```diff
@@ -161,6 +164,23 @@ int mbedtls_pk_write_pubkey( unsigned char **p, unsigned char *start,
+        if ( psa_export_public_key( *key_slot, start, buffer_size, &len )
+             != PSA_SUCCESS )
+        {
+            return MBEDTLS_ERR_PK_BAD_INPUT_DATA;
```
Minor: Brackets missing.

hanno-arm commented on library/pkwrite.c:
```diff
@@ -161,6 +164,23 @@ int mbedtls_pk_write_pubkey( unsigned char **p, unsigned char *start,
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+    if( mbedtls_pk_get_type( key ) == MBEDTLS_PK_OPAQUE )
+    {
+        size_t buffer_size = *p - start;
```
As a safe-guard, I'd prefer to check against `*p < start` here, and return `MBEDTLS_ERR_PK_BAD_INPUT_DATA` in this case.

hanno-arm commented on library/pkwrite.c:
```diff
@@ -161,6 +164,23 @@ int mbedtls_pk_write_pubkey( unsigned char **p, unsigned char *start,
+    if( mbedtls_pk_get_type( key ) == MBEDTLS_PK_OPAQUE )
+    {
+        size_t buffer_size = *p - start;
+        psa_key_slot_t* key_slot = ( psa_key_slot_t* ) key->pk_ctx;
```
(Very) minor: We usually write `(type) x` instead of `( type ) x`.

hanno-arm commented on tests/suites/test_suite_x509write.function:
```diff
@@ -95,6 +123,53 @@ exit:
+    if( cert_type != 0 )
+        TEST_ASSERT( mbedtls_x509write_csr_set_ns_cert_type( &req, cert_type ) == 0 );
+
+    ret = mbedtls_x509write_csr_pem( &req, buf, sizeof( buf ),
```
Minor: We should only provide `sizeof( buf ) - 1` bytes, to have space for the trailing `\0` byte.

hanno-arm commented on tests/suites/test_suite_x509write.function:
```diff
@@ -95,6 +123,53 @@ exit:
+
+    pem_len = strlen( (char *) buf );
+    buf[pem_len] = '\0';
+    TEST_ASSERT( x509_crt_verifycsr( buf, pem_len+1 ) == 0 );
```
Minor: Space between operands and `+`.

AndrzejKurek commented on tests/suites/test_suite_x509write.data:
```diff
@@ -42,6 +42,10 @@ Certificate Request check Server5 ECDSA, key_usage
 x509_csr_check:"data_files/server5.key":"data_files/server5.req.ku.sha1":MBEDTLS_MD_SHA1:MBEDTLS_X509_KU_DIGITAL_SIGNATURE | MBEDTLS_X509_KU_NON_REPUDIATION:0
 
+Certificate Request check opaque Server5 ECDSA, key_usage
+depends_on:MBEDTLS_SHA256_C:MBEDTLS_ECDSA_C:MBEDTLS_ECDSA_DETERMINISTIC:MBEDTLS_ECP_DP_SECP256R1_ENABLED:MBEDTLS_USE_PSA_CRYPTO
```
It's a leftover after copy-pasting the original CSR test. Removed, thanks!

AndrzejKurek commented:
Added all of the comments, awaiting the submodule PR.

hanno-arm commented on library/pkwrite.c:
```diff
@@ -167,8 +167,13 @@ int mbedtls_pk_write_pubkey( unsigned char **p, unsigned char *start,
+        if ( *p < start )
+            return( MBEDTLS_ERR_PK_BAD_INPUT_DATA );
+
+        buffer_size = *p - start;
```
Minor: I'm not sure if some compilers would complain about this implicit cast from signed to unsigned. Perhaps it's worth adding an `(size_t)` cast here.

AndrzejKurek commented:
@hanno-arm done!
By the way - this PR is tested by running test suites with and without the PSA define, I've also ran check-names, check-files, and check-doxy-blocks and ssl-opt.sh. I'll try to run all.sh if I have no other issues to resolve.
